src/bespokelabs/curator/request_processor/openai_batch_request_processor.py

```diff
@@ -743,18 +772,49 @@ async def track_already_submitted_batches(self):
 
-                    if request_file_name in self.tracker.unsubmitted_request_files:
-                        self.tracker.mark_as_submitted(request_file_name, batch_object, n_requests)
-                    else:
-                        # batch objects if not unsubmitted, should be downloaded
-                        assert batch_object.id in self.tracker.downloaded_batches
+                    # For each request file, we only want to keep the latest batch object.
+                    if (
+                        request_file_name not in existing_submitted_batches
+                        or existing_submitted_batches[request_file_name].created_at
+                        < batch_object.created_at
+                    ):
+                        existing_submitted_batches[request_file_name] = batch_object
+
+        for request_file_name, batch_object in existing_submitted_batches.items():
+
+            output_file_id = batch_object.output_file_id
+            if output_file_id is not None:
+                try:
+                    await self.client.files.retrieve(output_file_id)
+                except NotFoundError:
+                    logger.warning(
+                        f"Output file {output_file_id} exists in batch object but cannot be found "
+                        "in OpenAI storage. The file may have been deleted. Will resubmit this batch..."
+                    )
+                    continue
+
+            if request_file_name in self.tracker.unsubmitted_request_files:
+                self.tracker.mark_as_submitted(request_file_name, batch_object, n_requests)
+            else:
+                response_file = request_file_to_response_file(request_file_name, self.working_dir)
+                if not os.path.exists(response_file):
+                    raise ValueError(
+                        f"While processing {batch_object.id}, we found that its corresponding request_file_name {request_file_name} is "
+                        f"not in tracker.unsubmitted_request_files, but its corresponding response_file {response_file} does not exist. "
+                        f"This is an invalid state. \n"
+                        f"batch_object: {batch_object} \n"
+                        f"request_file_name: {request_file_name} \n"
+                        f"tracker.unsubmitted_request_files: {self.tracker.unsubmitted_request_files} \n"
+                        f"tracker.submitted_batches: {self.tracker.submitted_batches} \n"
+                        f"tracker.downloaded_batches: {self.tracker.downloaded_batches} \n"
+                    )
 
         if self.tracker.n_submitted_batches > 0:
             logger.info(
                 f"{self.tracker.n_submitted_batches:,} out of {self.tracker.n_total_batches - self.tracker.n_downloaded_batches:,} remaining batches are already submitted."
             )
 
-    def track_already_downloaded_batches(self):
+    async def track_already_downloaded_batches(self):
```

Comment:
what part of this is async? I believe we removed this before?

Reply:
ooops sorry, don't know what added it back for some reason.